<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Study bias in the protein network • MItools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Study bias in the protein network">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">MItools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.40</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/Full_interactomes_interspecies.html">Getting started: full interactomes and interspecies interactions</a>
    </li>
    <li>
      <a href="../../articles/articles/MItools_vs_igraph.html">MItools vs igraph: calculating degree of a protein</a>
    </li>
    <li>
      <a href="../../articles/articles/study_bias_doc.html">Study bias in the protein network</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Study bias in the protein network</h1>
                        <h4 class="author">Vitalii Kleshchevnikov</h4>
            
            <h4 class="date">11/01/2018</h4>
      
      
      <div class="hidden name"><code>study_bias_doc.Rmd</code></div>

    </div>

    
    
<p>This article and accompanying slides cover fundamentals of working with PPI data and the problem of study bias in protein interactions networks.</p>
<div id="outline" class="section level1">
<h1 class="hasAnchor">
<a href="#outline" class="anchor"></a>Outline</h1>
<p>This workshop covers the following topics:<br>
0. Brief intro into R Markdown<br>
1. The fundamentals of protein-protein interactions (PPI) and networks (lecture)<br>
2. Getting PPI data into R (from IntAct, using PSICQUIC)<br>
3. Finding publication (Pubmed) IDs for every protein<br>
4. Why should you care about the study bias in your network and how to evaluate it</p>
</div>
<div id="brief-intro-into-r-markdown" class="section level1">
<h1 class="hasAnchor">
<a href="#brief-intro-into-r-markdown" class="anchor"></a>0. Brief intro into R Markdown</h1>
<p>Learn more:<br><a href="http://rmarkdown.rstudio.com/" class="uri">http://rmarkdown.rstudio.com/</a></p>
<p>Cheat sheet:<br><a href="https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf" class="uri">https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf</a></p>
<p>Online course on Reproducible data analysis: <a href="https://www.coursera.org/learn/reproducible-research/lecture/5NzHN/r-markdown" class="uri">https://www.coursera.org/learn/reproducible-research/lecture/5NzHN/r-markdown</a></p>
<p>Slides:<br><a href="https://github.com/vitkl/PPI_biases_workshop/blob/master/PPI_networks_and_study_bias.pptx" class="uri">https://github.com/vitkl/PPI_biases_workshop/blob/master/PPI_networks_and_study_bias.pptx</a></p>
</div>
<div id="the-fundamentals-of-protein-protein-interactions-ppi-and-networks-lecture" class="section level1">
<h1 class="hasAnchor">
<a href="#the-fundamentals-of-protein-protein-interactions-ppi-and-networks-lecture" class="anchor"></a>1. The fundamentals of protein-protein interactions (PPI) and networks (lecture)</h1>
<p>The PowerPoint presentation for this lecture can be found in the project folder (PPI_networks_and_study_bias.pptx).</p>
</div>
<div id="getting-ppi-data-into-r-from-intact-using-psicquic" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-ppi-data-into-r-from-intact-using-psicquic" class="anchor"></a>2. Getting PPI data into R (from IntAct, using PSICQUIC)</h1>
<p>You can download protein-protein interaction data directly from IntAct or using PSICQUIC and PSICQUIC client implemented in PSICQUIC R package. Both ways are implemented in MItools package. Downloading data from IntAct gets you all data from all IMEx consortium databases. PSICQUIC route allows to query non-IMEx databases and construct complex search queries.<br>
As implemented in MItools, PSICQUIC route is slow to download (limit of 2500 interactions per second to lower the load on servers). IntAct route is faster but requires a large download (3GB, all data) and memory intensive processing locally.</p>
<p>First, let’s focus on the IntAct route. Later we will download molecular interactions from a large non-IMEx database called BioGRID.</p>
<div id="getting-ppi-data-into-r-from-intact" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-ppi-data-into-r-from-intact" class="anchor"></a>2.1 Getting PPI data into R from IntAct</h2>
<div id="load-full-human-human-interactome-and-count-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#load-full-human-human-interactome-and-count-interactions" class="anchor"></a>Load full human-human interactome and count interactions</h3>
<p>By species search is taxonomy-hierarchy aware.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find the date of the latest locally available release</span>
human =<span class="st"> </span><span class="kw"><a href="../../reference/fullInteractome.html">fullInteractome</a></span>(<span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">database =</span> <span class="st">"IntActFTP"</span>, <span class="co"># 9606 - human taxid</span>
                        <span class="dt">clean =</span> <span class="ot">TRUE</span>,
                        <span class="dt">protein_only =</span> <span class="ot">TRUE</span>,
                        <span class="dt">directory =</span> <span class="st">"./data_files/"</span>, <span class="co"># NULL to keep data files inside R library - default</span>
                        <span class="dt">releaseORdate =</span> <span class="ot">NULL</span>) <span class="co"># useful to keep track of the release date e.g. 2018Aug7, but for the first download set to NULL </span></code></pre></div>
<pre><code>## ... looking for the date of the latest IntAct release ...
## ... looking for the date of the latest IntAct release ...</code></pre>
<pre><code>## ... loading local copy ...</code></pre>
<pre><code>## Warning in fread(file_name, header = T, stringsAsFactors = F): Found
## and resolved improper quoting out-of-sample. First healed line 25383:
## &lt;&lt;uniprotkb:P16054 uniprotkb:Q05769 intact:EBI-298451 intact:EBI-298933|
## uniprotkb:Q543K3 psi-mi:kpce_mouse(display_long)|uniprotkb:Prkce(gene
## name)|psi-mi:Prkce(display_short)|uniprotkb:Pkce(gene name synonym)|
## uniprotkb:Pkcea(gene name synonym)|uniprotkb:nPKC-epsilon(gene name
## synonym) psi-mi:pgh2_mouse(display_long)|uniprotkb:Ptgs2(gene name)|
## psi-mi:Ptgs2(display_short)|uniprotkb:Cox-2(gene name synonym)|
## uniprotkb:Cox2(gene name synonym)|uniprotkb:Pghs-b(gene name synonym)|
## uniprotkb:Tis10(gene na&gt;&gt;. If the fields are not quoted (e.g. field
## separator does not appear within any field), try quote="" to avoid this
## warning.</code></pre>
</div>
<div id="load-all-human-viral-interactions-from-intact-biogrid-virhostnet" class="section level3">
<h3 class="hasAnchor">
<a href="#load-all-human-viral-interactions-from-intact-biogrid-virhostnet" class="anchor"></a>Load all human-viral interactions from IntAct, BioGRID, VirHostNet</h3>
<p>10239 is the top viral clade identifier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read all IntAct data (filter later)</span>
IntAct =<span class="st"> </span><span class="kw"><a href="../../reference/loadIntActFTP.html">loadIntActFTP</a></span>(<span class="dt">dir =</span> <span class="st">"./data_files/IntActRelease_2018Aug7/"</span>)</code></pre></div>
<pre><code>## ... looking for the date of the latest IntAct release ...</code></pre>
<pre><code>## ... loading local copy ...</code></pre>
<pre><code>## Warning in fread(file_name, header = T, stringsAsFactors = F): Found
## and resolved improper quoting out-of-sample. First healed line 25383:
## &lt;&lt;uniprotkb:P16054 uniprotkb:Q05769 intact:EBI-298451 intact:EBI-298933|
## uniprotkb:Q543K3 psi-mi:kpce_mouse(display_long)|uniprotkb:Prkce(gene
## name)|psi-mi:Prkce(display_short)|uniprotkb:Pkce(gene name synonym)|
## uniprotkb:Pkcea(gene name synonym)|uniprotkb:nPKC-epsilon(gene name
## synonym) psi-mi:pgh2_mouse(display_long)|uniprotkb:Ptgs2(gene name)|
## psi-mi:Ptgs2(display_short)|uniprotkb:Cox-2(gene name synonym)|
## uniprotkb:Cox2(gene name synonym)|uniprotkb:Pghs-b(gene name synonym)|
## uniprotkb:Tis10(gene na&gt;&gt;. If the fields are not quoted (e.g. field
## separator does not appear within any field), try quote="" to avoid this
## warning.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter for viral-human interaction</span>
human_viral =<span class="st"> </span><span class="kw"><a href="../../reference/interSpeciesInteractome.html">interSpeciesInteractome</a></span>(<span class="dt">MITABdata =</span> IntAct,
                                      <span class="dt">taxid1 =</span> <span class="dv">9606</span>,  <span class="dt">taxid2 =</span> <span class="dv">10239</span>,
                                      <span class="dt">database =</span> <span class="st">"IntActFTP"</span>,
                                      <span class="dt">clean =</span> <span class="ot">TRUE</span>, <span class="dt">protein_only =</span> <span class="ot">TRUE</span>, 
                                      <span class="dt">directory =</span> <span class="st">"./data_files/"</span>,
                        <span class="dt">releaseORdate =</span> <span class="st">"2018Aug7"</span>)</code></pre></div>
<pre><code>## Warning in fread(filename): Found and resolved improper quoting out-of-
## sample. First healed line 13569: &lt;&lt;1857845 Spo1virus "SPO1-like viruses";
## SP01-like viruses; SPO1-like viruses; Spounalikevirus Genus Viruses; dsDNA
## viruses, no RNA stage; Caudovirales; Myoviridae; Spounavirinae 857473 &gt;&gt;.
## If the fields are not quoted (e.g. field separator does not appear within
## any field), try quote="" to avoid this warning.</code></pre>
<pre><code>## Warning in fread(filename): Found and resolved improper quoting out-of-
## sample. First healed line 13569: &lt;&lt;1857845 Spo1virus "SPO1-like viruses";
## SP01-like viruses; SPO1-like viruses; Spounalikevirus Genus Viruses; dsDNA
## viruses, no RNA stage; Caudovirales; Myoviridae; Spounavirinae 857473 &gt;&gt;.
## If the fields are not quoted (e.g. field separator does not appear within
## any field), try quote="" to avoid this warning.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_viral)</code></pre></div>
<pre><code>## [1] 15003</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_viral, <span class="dt">taxid =</span> <span class="dv">9606</span>)</code></pre></div>
<pre><code>## [1] 4644</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_viral, <span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">inverse_filter =</span> T)</code></pre></div>
<pre><code>## [1] 894</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read all BioGRID data</span>
<span class="co">#download.file("https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.5.166/BIOGRID-ALL-3.5.166.mitab.zip",</span>
<span class="co">#              "./data_files/BIOGRID-ALL-3.5.166.mitab.zip")</span>
<span class="co">#unzip("./data_files/BIOGRID-ALL-3.5.166.mitab.zip")</span>
<span class="co">#biogrid = fread("./data_files/BIOGRID-ALL-3.5.166.mitab.txt")</span>
<span class="co">#setnames(biogrid, colnames(biogrid),</span>
<span class="co">#         colnames(IntAct$data)[1:15])</span>
<span class="co">#biogrid = list(data = biogrid, metadata = "full_biogrid_18_nov_2018")</span>
<span class="co">#class(biogrid) = "RAW_MItab25"</span>
<span class="co">#biogrid_vir = interSpeciesInteractome(taxid1 = 9606,  taxid2 = 10239, </span>
<span class="co">#                                      MITABdata = biogrid,</span>
<span class="co">#                                      clean = TRUE, protein_only = FALSE, </span>
<span class="co">#                                      directory = "./data_files/")</span>

<span class="co">#uniqueNinteractions(biogrid_vir)</span>
<span class="co">#uniqueNinteractors(biogrid_vir, taxid = 9606)</span>
<span class="co">#uniqueNinteractors(biogrid_vir, taxid = 9606, inverse_filter = T)</span>

<span class="co"># retrieve all VirHostNet data </span>
<span class="co"># which does not understand taxonomy hierarhy so interSpeciesInteractome() will not work</span>
VirHostNet_raw =<span class="st"> </span><span class="kw"><a href="../../reference/queryPSICQUICrlib.html">queryPSICQUICrlib</a></span>(<span class="dt">query =</span> <span class="st">"taxid:9606"</span>,
                                   <span class="dt">format =</span> <span class="st">"tab27"</span>, <span class="dt">database =</span> <span class="st">"VirHostNet"</span>,
                                   <span class="dt">directory =</span> <span class="st">"./data_files/"</span>,
                                   <span class="dt">releaseORdate =</span> <span class="ot">NULL</span>)</code></pre></div>
<pre><code>## downloading using PSICQUIC</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">VirHostNet =<span class="st"> </span><span class="kw"><a href="../../reference/interSpeciesInteractome.html">interSpeciesInteractome</a></span>(<span class="dt">taxid1 =</span> <span class="dv">9606</span>,  <span class="dt">taxid2 =</span> <span class="dv">10239</span>, 
                                     <span class="dt">MITABdata =</span> VirHostNet_raw,
                                     <span class="dt">clean =</span> <span class="ot">TRUE</span>, <span class="dt">protein_only =</span> <span class="ot">TRUE</span>, 
                                     <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)</code></pre></div>
<pre><code>## ... looking for the date of the latest IntAct release ...</code></pre>
<pre><code>## Warning in fread(filename): Found and resolved improper quoting out-of-
## sample. First healed line 13569: &lt;&lt;1857845 Spo1virus "SPO1-like viruses";
## SP01-like viruses; SPO1-like viruses; Spounalikevirus Genus Viruses; dsDNA
## viruses, no RNA stage; Caudovirales; Myoviridae; Spounavirinae 857473 &gt;&gt;.
## If the fields are not quoted (e.g. field separator does not appear within
## any field), try quote="" to avoid this warning.

## Warning in fread(filename): Found and resolved improper quoting out-of-
## sample. First healed line 13569: &lt;&lt;1857845 Spo1virus "SPO1-like viruses";
## SP01-like viruses; SPO1-like viruses; Spounalikevirus Genus Viruses; dsDNA
## viruses, no RNA stage; Caudovirales; Myoviridae; Spounavirinae 857473 &gt;&gt;.
## If the fields are not quoted (e.g. field separator does not appear within
## any field), try quote="" to avoid this warning.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter out human-human and viral-viral interactions </span>
VirHostNet<span class="op">$</span>data =<span class="st"> </span>VirHostNet<span class="op">$</span>data[<span class="op">!</span>Taxid_interactor_A <span class="op">==</span><span class="st"> </span>Taxid_interactor_B]
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(VirHostNet)</code></pre></div>
<pre><code>## [1] 23109</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(VirHostNet, <span class="dt">taxid =</span> <span class="dv">9606</span>)</code></pre></div>
<pre><code>## [1] 6216</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(VirHostNet, <span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">inverse_filter =</span> T)</code></pre></div>
<pre><code>## [1] 766</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Find which interactions are uniquely present in VirHostNet (not IntAct)</span>
VirHostNet<span class="op">$</span>data =<span class="st"> </span>VirHostNet<span class="op">$</span>data[<span class="op">!</span>pair_id <span class="op">%in%</span><span class="st"> </span>human_viral<span class="op">$</span>data<span class="op">$</span>pair_id]
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(VirHostNet)</code></pre></div>
<pre><code>## [1] 17822</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(VirHostNet, <span class="dt">taxid =</span> <span class="dv">9606</span>)</code></pre></div>
<pre><code>## [1] 5512</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(VirHostNet, <span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">inverse_filter =</span> T)</code></pre></div>
<pre><code>## [1] 679</code></pre>
</div>
<div id="load-all-human-mouse-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#load-all-human-mouse-interactions" class="anchor"></a>Load all human-mouse interactions</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter for viral-mouse interaction</span>
human_mouse =<span class="st"> </span><span class="kw"><a href="../../reference/interSpeciesInteractome.html">interSpeciesInteractome</a></span>(<span class="dt">taxid1 =</span> <span class="dv">9606</span>,  <span class="dt">taxid2 =</span> <span class="dv">10090</span>,
                                      <span class="dt">database =</span> <span class="st">"IntActFTP"</span>, <span class="dt">MITABdata =</span> IntAct,
                                      <span class="dt">clean =</span> <span class="ot">TRUE</span>, <span class="dt">protein_only =</span> <span class="ot">TRUE</span>,
                                      <span class="dt">directory =</span> <span class="st">"./data_files/"</span>,
                        <span class="dt">releaseORdate =</span> <span class="st">"2018Aug7"</span>)
<span class="co"># save RData file to save time processing this data if you want to proceed to later sections immediately</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(IntAct)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/save">save</a></span>(<span class="dt">list =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>(), <span class="dt">file =</span> <span class="st">"./RData_steps/study_bias_doc_1.RData"</span>)</code></pre></div>
<p>You can find out any other taxid if you know species name using UniProt taxonomy search: <a href="https://www.uniprot.org/taxonomy/" class="uri">https://www.uniprot.org/taxonomy/</a></p>
</div>
<div id="count-interactors-and-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#count-interactors-and-interactions" class="anchor"></a>Count interactors and interactions</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load previosly saved RData</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(<span class="st">"./RData_steps/study_bias_doc_1.RData"</span>)

<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human)</code></pre></div>
<pre><code>## [1] 196473</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human)</code></pre></div>
<pre><code>## [1] 21521</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_viral)</code></pre></div>
<pre><code>## [1] 15003</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_viral)</code></pre></div>
<pre><code>## [1] 5538</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_mouse)</code></pre></div>
<pre><code>## [1] 24960</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_mouse)</code></pre></div>
<pre><code>## [1] 10114</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># how many of these proteins are human?</span>
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_mouse, <span class="dt">taxid =</span> <span class="dv">9606</span>)</code></pre></div>
<pre><code>## [1] 7060</code></pre>
</div>
<div id="filter-human-human-data-by-detection-method-and-recount-interactors-and-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-human-human-data-by-detection-method-and-recount-interactors-and-interactions" class="anchor"></a>filter human-human data by detection method and recount interactors and interactions</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset two-hybrid interactions</span>
human_two_hybrid =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyMethod.html">subsetMITABbyMethod</a></span>(<span class="dt">MITABdata =</span> human,
                                       <span class="dt">Interaction_detection_methods =</span> <span class="st">"MI:0018"</span>)</code></pre></div>
<pre><code>## loading local copy of MI ontology</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_two_hybrid)</code></pre></div>
<pre><code>## [1] 83919</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_two_hybrid)</code></pre></div>
<pre><code>## [1] 14396</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset all interactions but two-hybrid</span>
human_NOT_two_hybrid =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyMethod.html">subsetMITABbyMethod</a></span>(<span class="dt">MITABdata =</span> human,
                                           <span class="dt">Interaction_detection_methods =</span> <span class="st">"MI:0018"</span>, <span class="dt">inverse_filter =</span> T)</code></pre></div>
<pre><code>## loading local copy of MI ontology</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_NOT_two_hybrid)</code></pre></div>
<pre><code>## [1] 116651</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_NOT_two_hybrid)</code></pre></div>
<pre><code>## [1] 16616</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset affinity purification - mass spectrometry interactions</span>
human_AP_MS =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyMethod.html">subsetMITABbyMethod</a></span>(<span class="dt">MITABdata =</span> human,
                                  <span class="dt">Interaction_detection_methods =</span> <span class="st">"MI:0004"</span>,  <span class="dt">Identification_method =</span> <span class="st">"MI:0433"</span>)</code></pre></div>
<pre><code>## loading local copy of MI ontology</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(human_AP_MS)</code></pre></div>
<pre><code>## [1] 69044</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(human_AP_MS)</code></pre></div>
<pre><code>## [1] 12465</code></pre>
</div>
<div id="filter-human-human-data-by-pmid-and-recount-interactors-and-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-human-human-data-by-pmid-and-recount-interactors-and-interactions" class="anchor"></a>filter human-human data by PMID and recount interactors and interactions</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset both published and unpublished Vidal group data</span>
Vidal_all =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyPMIDs.html">subsetMITABbyPMIDs</a></span>(<span class="dt">MITABdata =</span> human,
                               <span class="dt">PMIDs =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"25416956"</span>, <span class="st">"unassigned1304"</span>))
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(Vidal_all)</code></pre></div>
<pre><code>## [1] 59037</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(Vidal_all)</code></pre></div>
<pre><code>## [1] 9756</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset Mattias Mann 2015 paper data</span>
Mann =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyPMIDs.html">subsetMITABbyPMIDs</a></span>(<span class="dt">MITABdata =</span> human,
                          <span class="dt">PMIDs =</span> <span class="st">"26496610"</span>)
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(Mann)</code></pre></div>
<pre><code>## [1] 15589</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(Mann)</code></pre></div>
<pre><code>## [1] 4950</code></pre>
<p>You can get help and more details on these functions (for example, how to find which molecular ontology terms correspond to which methods): ?subsetMITABbyMethod</p>
</div>
<div id="find-interactions-between-components-of-the-mediator-complex-in-the-vidal-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#find-interactions-between-components-of-the-mediator-complex-in-the-vidal-dataset" class="anchor"></a>Find interactions between components of the mediator complex in the Vidal dataset</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mediator_complex_proteins =<span class="st"> </span><span class="kw">fread</span>(<span class="st">"http://www.uniprot.org/uniprot/?query=GO:0016592%20AND%20taxonomy:9606&amp;format=tab&amp;columns=id"</span>)
mediator_complex =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyID.html">subsetMITABbyID</a></span>(Vidal_all,
                                   <span class="dt">ID_seed =</span> mediator_complex_proteins<span class="op">$</span>Entry,
                                   <span class="dt">within_seed =</span> T, <span class="dt">only_seed2nonseed =</span> F)
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(mediator_complex)</code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(mediator_complex)</code></pre></div>
<pre><code>## [1] 7</code></pre>
</div>
<div id="find-interactions-of-the-components-of-the-mediator-complex-with-other-proteins" class="section level3">
<h3 class="hasAnchor">
<a href="#find-interactions-of-the-components-of-the-mediator-complex-with-other-proteins" class="anchor"></a>Find interactions of the components of the mediator complex with other proteins</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mediator_complex_interactions =<span class="st"> </span><span class="kw"><a href="../../reference/subsetMITABbyID.html">subsetMITABbyID</a></span>(Vidal_all,
                                                <span class="dt">ID_seed =</span> mediator_complex_proteins<span class="op">$</span>Entry,
                                                <span class="dt">within_seed =</span> F, <span class="dt">only_seed2nonseed =</span> T)
<span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractions</a></span>(mediator_complex_interactions)</code></pre></div>
<pre><code>## [1] 179</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/extractInteractors.html">uniqueNinteractors</a></span>(mediator_complex_interactions)</code></pre></div>
<pre><code>## [1] 179</code></pre>
</div>
</div>
<div id="getting-ppi-data-from-non-imex-databases-into-r-using-psicquic" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-ppi-data-from-non-imex-databases-into-r-using-psicquic" class="anchor"></a>2.2 Getting PPI data from non-IMEx databases into R using PSICQUIC</h2>
<p>Let’s retrieve all human-viral interactions from BioGrid database. But first we need to find correct database names.</p>
<div id="find-correct-database-names-" class="section level3">
<h3 class="hasAnchor">
<a href="#find-correct-database-names-" class="anchor"></a>Find correct database names.</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">client =<span class="st"> </span><span class="kw">PSICQUIC</span>()
<span class="kw">providers</span>(client)</code></pre></div>
<pre><code>##  [1] "BioGrid"           "bhf-ucl"           "ChEMBL"           
##  [4] "HPIDb"             "InnateDB"          "InnateDB-All"     
##  [7] "IntAct"            "IMEx"              "mentha"           
## [10] "MPIDB"             "iRefIndex"         "MatrixDB"         
## [13] "MINT"              "Reactome"          "Reactome-FIs"     
## [16] "EBI-GOA-miRNA"     "UniProt"           "MBInfo"           
## [19] "VirHostNet"        "BAR"               "EBI-GOA-nonIntAct"
## [22] "ZINC"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Don't run</span>
human_viral_BioGrid =<span class="st"> </span><span class="kw"><a href="../../reference/interSpeciesInteractome.html">interSpeciesInteractome</a></span>(<span class="dt">taxid1 =</span> <span class="dv">9606</span>,  <span class="dt">taxid2 =</span> <span class="dv">10239</span>,
                                              <span class="dt">database =</span> <span class="st">"BioGrid"</span>,
                                              <span class="dt">clean =</span> <span class="ot">TRUE</span>, <span class="dt">protein_only =</span> <span class="ot">TRUE</span>,
                                              <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)
human_BioGrid =<span class="st"> </span><span class="kw"><a href="../../reference/fullInteractome.html">fullInteractome</a></span>(<span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">database =</span> <span class="st">"BioGrid"</span>, <span class="co"># 9606 - human taxid</span>
                        <span class="dt">clean =</span> <span class="ot">TRUE</span>,
                        <span class="dt">protein_only =</span> <span class="ot">TRUE</span>,
                        <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)</code></pre></div>
</div>
</div>
<div id="advanced-search" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-search" class="anchor"></a>2.3 Advanced search</h2>
<p>This can be useful in you need interactions for a small number of proteins or if you want to query non-IMEx databases. Note that queryPSICQUIC doesn’t keep track of database version data, while queryPSICQUICrlib does.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Query for interactions of bacterial RNA polymerase sigma factor SigA identified using two-hybrid methods in all imex databases</span>
<span class="kw"><a href="../../reference/queryPSICQUIC.html">queryPSICQUIC</a></span>(<span class="dt">query =</span> <span class="st">"id:P74565 AND detmethod:</span><span class="ch">\"</span><span class="st">MI:0018</span><span class="ch">\"</span><span class="st">"</span>,
              <span class="dt">format =</span> <span class="st">"tab27"</span>,
              <span class="dt">database =</span> <span class="st">"imex"</span>,
              <span class="dt">file =</span> <span class="st">"./data_files/P74565_2H_interactions_imex_tab27.tsv"</span>)</code></pre></div>
<pre><code>## Warning in if (N_interactions &gt; 0) {: the condition has length &gt; 1 and only
## the first element will be used</code></pre>
<pre><code>## Warning in N_SPECIES_ID_interactome[indices] &lt;- N_interactions: number of
## items to replace is not a multiple of replacement length</code></pre>
<pre><code>##                                     query
##  1: id:P74565 AND detmethod:%22MI:0018%22
##  2: id:P74565 AND detmethod:%22MI:0018%22
##  3: id:P74565 AND detmethod:%22MI:0018%22
##  4: id:P74565 AND detmethod:%22MI:0018%22
##  5: id:P74565 AND detmethod:%22MI:0018%22
##  6: id:P74565 AND detmethod:%22MI:0018%22
##  7: id:P74565 AND detmethod:%22MI:0018%22
##  8: id:P74565 AND detmethod:%22MI:0018%22
##  9: id:P74565 AND detmethod:%22MI:0018%22
## 10: id:P74565 AND detmethod:%22MI:0018%22
## 11: id:P74565 AND detmethod:%22MI:0018%22
##                                                   file format
##  1: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  2: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  3: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  4: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  5: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  6: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  7: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  8: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##  9: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
## 10: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
## 11: ./data_files/P74565_2H_interactions_imex_tab27.tsv  tab27
##     all.databases n.interactions.in.database database.not.active
##  1:        IntAct                         44                    
##  2:          MINT                          0                    
##  3:       bhf-ucl                          0                    
##  4:         MPIDB                          0                    
##  5:      MatrixDB                     &lt;html&gt;                    
##  6:         HPIDb                          0                    
##  7:      I2D-IMEx                                       I2D-IMEx
##  8: InnateDB-IMEx                                  InnateDB-IMEx
##  9:        MolCon                                         MolCon
## 10:       UniProt                          0                    
## 11:        MBInfo                          0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Query for interactions of sigma factor SigA identified using two-hybrid methods in mentha (a database that aggregates data from all primary databases, but does no interaction predition)</span>
<span class="kw"><a href="../../reference/queryPSICQUIC.html">queryPSICQUIC</a></span>(<span class="dt">query =</span> <span class="st">"id:P74565 AND detmethod:</span><span class="ch">\"</span><span class="st">MI:0018</span><span class="ch">\"</span><span class="st">"</span>,
              <span class="dt">format =</span> <span class="st">"tab25"</span>,
              <span class="dt">database =</span> <span class="st">"mentha"</span>,
              <span class="dt">file =</span> <span class="st">"./data_files/P74565_2H_interactions_mentha_tab25.tsv"</span>)</code></pre></div>
<pre><code>##                                    query
## 1: id:P74565 AND detmethod:%22MI:0018%22
##                                                    file format
## 1: ./data_files/P74565_2H_interactions_mentha_tab25.tsv  tab25
##    all.databases n.interactions.in.database database.not.active
## 1:        mentha                         44</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Query for interactions of sigma factor SigA in mentha</span>
<span class="kw"><a href="../../reference/queryPSICQUIC.html">queryPSICQUIC</a></span>(<span class="dt">query =</span> <span class="st">"id:P74565"</span>,
              <span class="dt">format =</span> <span class="st">"tab25"</span>,
              <span class="dt">database =</span> <span class="st">"mentha"</span>,
              <span class="dt">file =</span> <span class="st">"./data_files/P74565_2H_interactions_mentha_tab25.tsv"</span>)</code></pre></div>
<pre><code>##        query                                                 file format
## 1: id:P74565 ./data_files/P74565_2H_interactions_mentha_tab25.tsv  tab25
##    all.databases n.interactions.in.database database.not.active
## 1:        mentha                         44</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Retrieve interaction of any proteins encoded by Beta-adrenergic receptor kinase 1 gene (Entrez GeneID 156) from BioGRID (which recognises only this type of ID)</span>
<span class="kw"><a href="../../reference/queryPSICQUIC.html">queryPSICQUIC</a></span>(<span class="dt">query =</span> <span class="st">"id:156"</span>,
              <span class="dt">format =</span> <span class="st">"tab25"</span>,
              <span class="dt">database =</span> <span class="st">"BioGrid"</span>,
              <span class="dt">file =</span> <span class="st">"./data_files/entrezgene156_interactions_BioGrid_tab25.tsv"</span>)</code></pre></div>
<pre><code>##     query                                                      file format
## 1: id:156 ./data_files/entrezgene156_interactions_BioGrid_tab25.tsv  tab25
##    all.databases n.interactions.in.database database.not.active
## 1:       BioGrid                         82</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The function return the report of how many interaction were found in each database, not the data itself. Reading data into R.</span>
<span class="kw">fread</span>(<span class="st">"./data_files/entrezgene156_interactions_BioGrid_tab25.tsv"</span>, <span class="dt">header =</span> T, <span class="dt">stringsAsFactors =</span> F)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>##                            V1                         V2
## 1:  entrez gene/locuslink:156 entrez gene/locuslink:5579
## 2:  entrez gene/locuslink:156 entrez gene/locuslink:5148
## 3:  entrez gene/locuslink:156 entrez gene/locuslink:6714
## 4: entrez gene/locuslink:6714  entrez gene/locuslink:156
## 5: entrez gene/locuslink:5148  entrez gene/locuslink:156
##                                                                             V3
## 1:                                 biogrid:106665|entrez gene/locuslink:ADRBK1
## 2:                                 biogrid:106665|entrez gene/locuslink:ADRBK1
## 3:                                 biogrid:106665|entrez gene/locuslink:ADRBK1
## 4: biogrid:112592|entrez gene/locuslink:SRC|entrez gene/locuslink:RP5-823N20.1
## 5:                                  biogrid:111174|entrez gene/locuslink:PDE6G
##                                                                             V4
## 1:                                  biogrid:111565|entrez gene/locuslink:PRKCB
## 2:                                  biogrid:111174|entrez gene/locuslink:PDE6G
## 3: biogrid:112592|entrez gene/locuslink:SRC|entrez gene/locuslink:RP5-823N20.1
## 4:                                 biogrid:106665|entrez gene/locuslink:ADRBK1
## 5:                                 biogrid:106665|entrez gene/locuslink:ADRBK1
##                                                                                                                                                                                            V5
## 1:                                            entrez gene/locuslink:BARK1(gene name synonym)|entrez gene/locuslink:BETA-ARK1(gene name synonym)|entrez gene/locuslink:GRK2(gene name synonym)
## 2:                                            entrez gene/locuslink:BARK1(gene name synonym)|entrez gene/locuslink:BETA-ARK1(gene name synonym)|entrez gene/locuslink:GRK2(gene name synonym)
## 3:                                            entrez gene/locuslink:BARK1(gene name synonym)|entrez gene/locuslink:BETA-ARK1(gene name synonym)|entrez gene/locuslink:GRK2(gene name synonym)
## 4: entrez gene/locuslink:ASV(gene name synonym)|entrez gene/locuslink:SRC1(gene name synonym)|entrez gene/locuslink:c-SRC(gene name synonym)|entrez gene/locuslink:p60-Src(gene name synonym)
## 5:                                                                                                entrez gene/locuslink:PDEG(gene name synonym)|entrez gene/locuslink:RP57(gene name synonym)
##                                                                                                                                                                                                 V6
## 1: entrez gene/locuslink:PKC-beta(gene name synonym)|entrez gene/locuslink:PKCB(gene name synonym)|entrez gene/locuslink:PRKCB1(gene name synonym)|entrez gene/locuslink:PRKCB2(gene name synonym)
## 2:                                                                                                     entrez gene/locuslink:PDEG(gene name synonym)|entrez gene/locuslink:RP57(gene name synonym)
## 3:      entrez gene/locuslink:ASV(gene name synonym)|entrez gene/locuslink:SRC1(gene name synonym)|entrez gene/locuslink:c-SRC(gene name synonym)|entrez gene/locuslink:p60-Src(gene name synonym)
## 4:                                                 entrez gene/locuslink:BARK1(gene name synonym)|entrez gene/locuslink:BETA-ARK1(gene name synonym)|entrez gene/locuslink:GRK2(gene name synonym)
## 5:                                                 entrez gene/locuslink:BARK1(gene name synonym)|entrez gene/locuslink:BETA-ARK1(gene name synonym)|entrez gene/locuslink:GRK2(gene name synonym)
##                                                    V7             V8
## 1: psi-mi:MI:0004(affinity chromatography technology) Yang XL (2003)
## 2: psi-mi:MI:0004(affinity chromatography technology)  Wan KF (2003)
## 3: psi-mi:MI:0004(affinity chromatography technology)  Wan KF (2003)
## 4: psi-mi:MI:0004(affinity chromatography technology)  Wan KF (2003)
## 5: psi-mi:MI:0004(affinity chromatography technology)  Wan KF (2003)
##                 V9        V10        V11
## 1: pubmed:12679936 taxid:9606 taxid:9606
## 2: pubmed:12624098 taxid:9606 taxid:9606
## 3: pubmed:12624098 taxid:9606 taxid:9606
## 4: pubmed:12624098 taxid:9606 taxid:9606
## 5: pubmed:12624098 taxid:9606 taxid:9606
##                                     V12                     V13
## 1: psi-mi:MI:0915(physical association) psi-mi:MI:0463(biogrid)
## 2: psi-mi:MI:0915(physical association) psi-mi:MI:0463(biogrid)
## 3: psi-mi:MI:0915(physical association) psi-mi:MI:0463(biogrid)
## 4: psi-mi:MI:0915(physical association) psi-mi:MI:0463(biogrid)
## 5: psi-mi:MI:0915(physical association) psi-mi:MI:0463(biogrid)
##               V14 V15
## 1: biogrid:259364   -
## 2: biogrid:259971   -
## 3: biogrid:259972   -
## 4: biogrid:259974   -
## 5: biogrid:259977   -</code></pre>
<p>All the same operations can be done using function queryPSICQUICrlib but with the convienience of automatic tracking of database release date and the exact query text. This function also return the data in object of class RAW_MItab that after cleaned make data ready for use with other tools in the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BioGrid_<span class="dv">156</span> =<span class="st"> </span><span class="kw"><a href="../../reference/queryPSICQUICrlib.html">queryPSICQUICrlib</a></span>(<span class="dt">query =</span> <span class="st">"id:156"</span>,
                                <span class="dt">format =</span> <span class="st">"tab25"</span>,
                                <span class="dt">database =</span> <span class="st">"BioGrid"</span>,
                                <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)</code></pre></div>
<pre><code>## downloading using PSICQUIC</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The same protein, but only two-hybrid interactions</span>
BioGrid_156_2H =<span class="st"> </span><span class="kw"><a href="../../reference/queryPSICQUICrlib.html">queryPSICQUICrlib</a></span>(<span class="dt">query =</span> <span class="st">"id:156 AND detmethod:</span><span class="ch">\"</span><span class="st">MI:0018</span><span class="ch">\"</span><span class="st">"</span>,
                                   <span class="dt">format =</span> <span class="st">"tab25"</span>,
                                   <span class="dt">database =</span> <span class="st">"BioGrid"</span>,
                                   <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)</code></pre></div>
<pre><code>## downloading using PSICQUIC</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The data returned by queryPSICQUICrlib constains auxillary information that is not necessary for most analysis. Let's clean the data.</span>
<span class="kw"><a href="../../reference/cleanMITAB.html">cleanMITAB</a></span>(BioGrid_<span class="dv">156</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>## $data
##     IDs_interactor_A IDs_interactor_B interactor_IDs_databases_A
##  1:              156             5579      entrez gene/locuslink
##  2:              156             5148      entrez gene/locuslink
##  3:              156             6714      entrez gene/locuslink
##  4:              156             2776      entrez gene/locuslink
##  5:              156             2769      entrez gene/locuslink
##  6:              156             5159      entrez gene/locuslink
##  7:             1460              156      entrez gene/locuslink
##  8:              156            25489      entrez gene/locuslink
##  9:              156             9815      entrez gene/locuslink
## 10:              156            28964      entrez gene/locuslink
## 11:              156             6010      entrez gene/locuslink
## 12:              156              857      entrez gene/locuslink
## 13:             1233              156      entrez gene/locuslink
## 14:              156             1910      entrez gene/locuslink
## 15:              156              185      entrez gene/locuslink
## 16:              156             2492      entrez gene/locuslink
## 17:              154              156      entrez gene/locuslink
## 18:              156            28964      entrez gene/locuslink
## 19:              156             5082      entrez gene/locuslink
## 20:              156             5132      entrez gene/locuslink
## 21:              156             6622      entrez gene/locuslink
## 22:              156             4193      entrez gene/locuslink
## 23:              156              408      entrez gene/locuslink
## 24:              156             2782      entrez gene/locuslink
## 25:              156             2782      entrez gene/locuslink
## 26:              156            30818      entrez gene/locuslink
## 27:              156            26260      entrez gene/locuslink
## 28:              154              156      entrez gene/locuslink
## 29:              156             5037      entrez gene/locuslink
## 30:              156           223864      entrez gene/locuslink
## 31:              156             4193      entrez gene/locuslink
## 32:              156             2776      entrez gene/locuslink
## 33:              156             1656      entrez gene/locuslink
## 34:              156            54919      entrez gene/locuslink
## 35:            10979              156      entrez gene/locuslink
## 36:            10914              156      entrez gene/locuslink
## 37:              156            23398      entrez gene/locuslink
## 38:            11212              156      entrez gene/locuslink
## 39:              156             5792      entrez gene/locuslink
## 40:              156             2911      entrez gene/locuslink
## 41:              156             1969      entrez gene/locuslink
## 42:              156             5923      entrez gene/locuslink
## 43:              156             3320      entrez gene/locuslink
## 44:              156             3326      entrez gene/locuslink
## 45:              156              157      entrez gene/locuslink
## 46:              156               31      entrez gene/locuslink
## 47:              156             4343      entrez gene/locuslink
## 48:              156             4193      entrez gene/locuslink
## 49:              156             2840      entrez gene/locuslink
## 50:              156             2783      entrez gene/locuslink
## 51:              156            16554      entrez gene/locuslink
## 52:              156            53970      entrez gene/locuslink
## 53:              156           238799      entrez gene/locuslink
## 54:              156             6257      entrez gene/locuslink
## 55:              156             9128      entrez gene/locuslink
## 56:              156             9646      entrez gene/locuslink
## 57:              156            57606      entrez gene/locuslink
## 58:              156            79731      entrez gene/locuslink
## 59:            10013              156      entrez gene/locuslink
## 60:           814753           816413      entrez gene/locuslink
## 61:           814753           829142      entrez gene/locuslink
## 62:           814753           829480      entrez gene/locuslink
## 63:              156            26051      entrez gene/locuslink
## 64:              156              157      entrez gene/locuslink
## 65:              156            57533      entrez gene/locuslink
##     IDs_interactor_A IDs_interactor_B interactor_IDs_databases_A
##     interactor_IDs_databases_B Taxid_interactor_A Taxid_interactor_B
##  1:      entrez gene/locuslink               9606               9606
##  2:      entrez gene/locuslink               9606               9606
##  3:      entrez gene/locuslink               9606               9606
##  4:      entrez gene/locuslink               9606               9606
##  5:      entrez gene/locuslink               9606               9606
##  6:      entrez gene/locuslink               9606               9606
##  7:      entrez gene/locuslink               9606               9606
##  8:      entrez gene/locuslink               9606              10116
##  9:      entrez gene/locuslink               9606               9606
## 10:      entrez gene/locuslink               9606               9606
## 11:      entrez gene/locuslink               9606               9606
## 12:      entrez gene/locuslink               9606               9606
## 13:      entrez gene/locuslink               9606               9606
## 14:      entrez gene/locuslink               9606               9606
## 15:      entrez gene/locuslink               9606               9606
## 16:      entrez gene/locuslink               9606               9606
## 17:      entrez gene/locuslink               9606               9606
## 18:      entrez gene/locuslink               9606               9606
## 19:      entrez gene/locuslink               9606               9606
## 20:      entrez gene/locuslink               9606               9606
## 21:      entrez gene/locuslink               9606               9606
## 22:      entrez gene/locuslink               9606               9606
## 23:      entrez gene/locuslink               9606               9606
## 24:      entrez gene/locuslink               9606               9606
## 25:      entrez gene/locuslink               9606               9606
## 26:      entrez gene/locuslink               9606               9606
## 27:      entrez gene/locuslink               9606               9606
## 28:      entrez gene/locuslink               9606               9606
## 29:      entrez gene/locuslink               9606               9606
## 30:      entrez gene/locuslink               9606              10090
## 31:      entrez gene/locuslink               9606               9606
## 32:      entrez gene/locuslink               9606               9606
## 33:      entrez gene/locuslink               9606               9606
## 34:      entrez gene/locuslink               9606               9606
## 35:      entrez gene/locuslink               9606               9606
## 36:      entrez gene/locuslink               9606               9606
## 37:      entrez gene/locuslink               9606               9606
## 38:      entrez gene/locuslink               9606               9606
## 39:      entrez gene/locuslink               9606               9606
## 40:      entrez gene/locuslink               9606               9606
## 41:      entrez gene/locuslink               9606               9606
## 42:      entrez gene/locuslink               9606               9606
## 43:      entrez gene/locuslink               9606               9606
## 44:      entrez gene/locuslink               9606               9606
## 45:      entrez gene/locuslink               9606               9606
## 46:      entrez gene/locuslink               9606               9606
## 47:      entrez gene/locuslink               9606               9606
## 48:      entrez gene/locuslink               9606               9606
## 49:      entrez gene/locuslink               9606               9606
## 50:      entrez gene/locuslink               9606               9606
## 51:      entrez gene/locuslink               9606              10090
## 52:      entrez gene/locuslink               9606              10090
## 53:      entrez gene/locuslink               9606              10090
## 54:      entrez gene/locuslink               9606               9606
## 55:      entrez gene/locuslink               9606               9606
## 56:      entrez gene/locuslink               9606               9606
## 57:      entrez gene/locuslink               9606               9606
## 58:      entrez gene/locuslink               9606               9606
## 59:      entrez gene/locuslink               9606               9606
## 60:      entrez gene/locuslink               3702               3702
## 61:      entrez gene/locuslink               3702               3702
## 62:      entrez gene/locuslink               3702               3702
## 63:      entrez gene/locuslink               9606               9606
## 64:      entrez gene/locuslink               9606               9606
## 65:      entrez gene/locuslink               9606               9606
##     interactor_IDs_databases_B Taxid_interactor_A Taxid_interactor_B
##     Publication_Identifiers Confidence_values       pair_id
##  1:                12679936                NA      156|5579
##  2:                12624098                NA      156|5148
##  3:                12624098                NA      156|6714
##  4:                12885252                NA      156|2776
##  5:                12885252                NA      156|2769
##  6:                12381737                NA      156|5159
##  7:                21900206                NA      1460|156
##  8:                19953087                NA     156|25489
##  9:                10896954                NA      156|9815
## 10:                10896954                NA     156|28964
## 11:                10722682                NA      156|6010
## 12:                10085129                NA       156|857
## 13:                10754297                NA      1233|156
## 14:                 9211925                NA      156|1910
## 15:                 9211925                NA       156|185
## 16:                10379886                NA      156|2492
## 17:                11309381                NA       154|156
## 18:                 9826657                NA     156|28964
## 19:                10884381                NA      156|5082
## 20:                10884381                NA      156|5132
## 21:                17893145                NA      156|6622
## 22:                21081496                NA      156|4193
## 23:                21081496                NA       156|408
## 24:                17548351                NA      156|2782
## 25:                10744734                NA      156|2782
## 26:                17102134                NA     156|30818
## 27:                23940030                NA     156|26260
## 28:                14654844                NA       154|156
## 29:                14654844                NA      156|5037
## 30:                20861385                NA    156|223864
## 31:                17006543                NA      156|4193
## 32:                21940795                NA      156|2776
## 33:                26344197                NA      156|1656
## 34:                26344197                NA     156|54919
## 35:                26344197                NA     10979|156
## 36:                26344197                NA     10914|156
## 37:                26344197                NA     156|23398
## 38:                26344197                NA     11212|156
## 39:                26344197                NA      156|5792
## 40:                16091361                NA      156|2911
## 41:                21988832                NA      156|1969
## 42:                21988832                NA      156|5923
## 43:                23455922                NA      156|3320
## 44:                23455922                NA      156|3326
## 45:                23455922                NA       156|157
## 46:                23455922                NA        156|31
## 47:                22658674                NA      156|4343
## 48:                26228571                NA      156|4193
## 49:                26228571                NA      156|2840
## 50:                25982117                NA      156|2783
## 51:                26496610                NA     156|16554
## 52:                26496610                NA     156|53970
## 53:                26496610                NA    156|238799
## 54:                26496610                NA      156|6257
## 55:                26496610                NA      156|9128
## 56:                26496610                NA      156|9646
## 57:                26496610                NA     156|57606
## 58:                26344197                NA     156|79731
## 59:                22193721                NA     10013|156
## 60:                26071421                NA 814753|816413
## 61:                26071421                NA 814753|829142
## 62:                26071421                NA 814753|829480
## 63:                28514442                NA     156|26051
## 64:                28514442                NA       156|157
## 65:                28514442                NA     156|57533
##     Publication_Identifiers Confidence_values       pair_id
## 
## $metadata
##     query                                                            file
## 1: id:156 ./data_files/DownloadDate_20190401/query_id.156tab25BioGrid.tsv
##    format all.databases n.interactions.in.database database.not.active
## 1:  tab25       BioGrid                         82                    
##    date_time
## 1: 2019Mar23
## 
## $&lt;NA&gt;
## NULL
## 
## $&lt;NA&gt;
## NULL
## 
## $&lt;NA&gt;
## NULL</code></pre>
</div>
</div>
<div id="finding-publication-pubmed-ids-for-every-protein" class="section level1">
<h1 class="hasAnchor">
<a href="#finding-publication-pubmed-ids-for-every-protein" class="anchor"></a>3. Finding publication (Pubmed) IDs for every protein</h1>
<p>There are many ways to find publications for a given protein (all proteins). We will discuss 3 and focus on one (UniProtKB references).</p>
<ol style="list-style-type: decimal">
<li>UniProtKB references. Details: <a href="https://www.uniprot.org/help/publications_section" class="uri">https://www.uniprot.org/help/publications_section</a><br>
Quote:<br>
“The set of publications fully curated in UniProtKB/Swiss-Prot and publications imported in UniProtKB/TrEMBL is complemented by additional publications that have been computationally mapped from other resources to UniProtKB entries.”</li>
</ol>
<p><a href="https://www.uniprot.org/help/publications_section#additional_bibliography" class="uri">https://www.uniprot.org/help/publications_section#additional_bibliography</a>:<br>
“As a comprehensive and high-quality resource of protein sequence and functional information, UniProtKB strives to provide comprehensive literature citations associated with protein sequences and their characterization. Currently about 2 thirds of the UniProtKB PubMed citations are found in UniProtKB/Swiss-Prot, as a result of active integration in the course of manual curation.</p>
<p>In order to keep up with the explosive growth of literature and to give our users access to additional publications, we decided to integrate additional sources of literature from other annotated databases into UniProtKB. For this purpose we selected a number external databases, e.g. Entrez Gene (GeneRIFs), SGD, MGI, GAD and PDB, and extracted citations that were mapped to UniProtKB entries. This additional protein bibliography information helps our users to better explore the existing knowledge of their proteins of interest."</p>
<ol start="2" style="list-style-type: decimal">
<li><p>Europe PMC text-mining efforts yeld multiple identifier-publication associations. Details can be found here: <a href="https://europepmc.org/downloads" class="uri">https://europepmc.org/downloads</a>. UniProtKB to pubmedID mapping can be downloaded from ftp: <a href="ftp://ftp.ebi.ac.uk/pub/databases/pmc/TextMinedTerms/" class="uri">ftp://ftp.ebi.ac.uk/pub/databases/pmc/TextMinedTerms/</a></p></li>
<li><p>NCBI portal: references for entrez geneID. Similar to UniProtKB references, but different ID, documentation about how these links are maintained is provided here: <a href="https://www.ncbi.nlm.nih.gov/entrez/query/static/entrezlinks.html#gene" class="uri">https://www.ncbi.nlm.nih.gov/entrez/query/static/entrezlinks.html#gene</a>. GeneID to pubmedID mapping can be downloaded from NCBI ftp: <a href="ftp://ftp.ncbi.nih.gov/gene/DATA/gene2pubmed.gz" class="uri">ftp://ftp.ncbi.nih.gov/gene/DATA/gene2pubmed.gz</a></p></li>
</ol>
<div id="retrive-uniprotkb-references-using-uniprot-rest-api" class="section level2">
<h2 class="hasAnchor">
<a href="#retrive-uniprotkb-references-using-uniprot-rest-api" class="anchor"></a>3.1 Retrive UniProtKB references using Uniprot REST API</h2>
<p>Let’s download the list of NCBI PubMed identifiers (PubMed IDs) associated with the UniProtKB entry (human proteins) and computationally mapped to the UniProtKB entry (human proteins).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># retrieve human interactions (taxonomy:9606) from SwissProt, a manually reviewed part or the UniProt (reviewed:yes)</span>
url =<span class="st"> "https://www.uniprot.org/uniprot/?query=taxonomy:9606&amp;compress=yes&amp;format=tab&amp;columns=id,citation,citationmapping"</span>
filename.gz =<span class="st"> "./data_files/uniprot2pubmed_human.tsv.gz"</span>
filename =<span class="st"> "./data_files/uniprot2pubmed_human.tsv"</span>
<span class="cf">if</span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files">file.exists</a></span>(filename.gz)) <span class="kw">download</span>(url, filename.gz)
R.utils<span class="op">::</span><span class="kw">gunzip</span>(filename.gz, filename, <span class="dt">remove =</span> F, <span class="dt">overwrite =</span> T)

<span class="co"># Read into R</span>
uniprot2pubmed_human =<span class="st"> </span><span class="kw">fread</span>(filename, <span class="dt">header =</span> T, <span class="dt">stringsAsFactors =</span> F)

<span class="co"># We need to do some processing</span>
## 1. Merge manually associated and computationally mapped publications 
uniprot2pubmed_human[, PMIDs <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">`</span><span class="dt">PubMed ID</span><span class="st">`</span>,<span class="st">";"</span>,<span class="st">`</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span>)]
uniprot2pubmed_human[<span class="st">`</span><span class="dt">PubMed ID</span><span class="st">`</span> <span class="op">==</span><span class="st"> ""</span> <span class="op">&amp;</span><span class="st"> `</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span> <span class="op">==</span><span class="st"> ""</span>, PMIDs <span class="op">:</span><span class="er">=</span><span class="st"> ""</span>]
uniprot2pubmed_human[<span class="st">`</span><span class="dt">PubMed ID</span><span class="st">`</span> <span class="op">==</span><span class="st"> ""</span> <span class="op">&amp;</span><span class="st"> `</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span> <span class="op">!=</span><span class="st"> ""</span>, PMIDs <span class="op">:</span><span class="er">=</span><span class="st"> `</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span>]
uniprot2pubmed_human[<span class="st">`</span><span class="dt">PubMed ID</span><span class="st">`</span> <span class="op">!=</span><span class="st"> ""</span> <span class="op">&amp;</span><span class="st"> `</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span> <span class="op">==</span><span class="st"> ""</span>, PMIDs <span class="op">:</span><span class="er">=</span><span class="st"> `</span><span class="dt">PubMed ID</span><span class="st">`</span>]
uniprot2pubmed_human<span class="op">$</span><span class="st">`</span><span class="dt">PubMed ID</span><span class="st">`</span> =<span class="st"> </span><span class="ot">NULL</span>
uniprot2pubmed_human<span class="op">$</span><span class="st">`</span><span class="dt">Mapped PubMed ID</span><span class="st">`</span> =<span class="st"> </span><span class="ot">NULL</span>
## 2. Split concatenated PubMed IDs and count their number per protein
uniprot2pubmed_human =<span class="st"> </span>uniprot2pubmed_human[, .(<span class="dt">N_PMIDs =</span> <span class="kw">uniqueN</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(<span class="kw">tstrsplit</span>(PMIDs,<span class="st">";"</span>)))), by =<span class="st"> </span>Entry]

<span class="co"># remove objects necessary for the next part, save</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(<span class="dt">list =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>()[<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>() <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"human"</span>, <span class="st">"uniprot2pubmed_human"</span>, <span class="st">"Vidal_all"</span>, <span class="st">"Mann"</span>, <span class="st">"human_two_hybrid"</span>, <span class="st">"human_AP_MS"</span>)])
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/save">save</a></span>(<span class="dt">list =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>(), <span class="dt">file =</span> <span class="st">"./RData_steps/study_bias_doc_2.RData"</span>)</code></pre></div>
</div>
</div>
<div id="evaluating-bias-of-protein-popularity-in-the-literature" class="section level1">
<h1 class="hasAnchor">
<a href="#evaluating-bias-of-protein-popularity-in-the-literature" class="anchor"></a>4. Evaluating bias of protein popularity in the literature</h1>
<div id="why-should-you-care-about-the-study-bias-in-your-network-and-how-to-evaluate-it" class="section level3">
<h3 class="hasAnchor">
<a href="#why-should-you-care-about-the-study-bias-in-your-network-and-how-to-evaluate-it" class="anchor"></a>Why should you care about the study bias in your network and how to evaluate it</h3>
<p>Back to the presentation, slide 61. Study bias is the phenomenon that causes certain properties of proteins to appear biologically meaningful but actually confounded with how well studies proteins are overall. Study bias may cause certain proteins to appear as hubs in the protein interaction network. This may lead to false conclusions regarding the importance of the protein in the network. Therefore, it is important to estimate and if possible correct the effect of study bias on the property of your interest.</p>
</div>
<div id="first-we-need-to-bin-proteins-by-how-often-they-are-mentioned-in-the-literature" class="section level3">
<h3 class="hasAnchor">
<a href="#first-we-need-to-bin-proteins-by-how-often-they-are-mentioned-in-the-literature" class="anchor"></a>First, we need to bin proteins by how often they are mentioned in the literature</h3>
<p>Let’s use the number of UniProt reference as a proxy for how well studied the protein is. We will bin all proteins that have interactions in our network by the number of publications.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(<span class="st">"./RData_steps/study_bias_doc_2.RData"</span>)
uniprot2pubmed_human =<span class="st"> </span>uniprot2pubmed_human[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(N_PMIDs, <span class="dt">decreasing =</span> T)]
uniprot2pubmed_human_w =<span class="st"> </span>uniprot2pubmed_human[Entry <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="../../reference/extractInteractors.html">extractInteractors</a></span>(human)]

<span class="co"># Most proteins have very low number of publications:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(uniprot2pubmed_human_w<span class="op">$</span>N_PMIDs), <span class="dt">xlab =</span> <span class="st">"log10(number of publications) per protein"</span>)</code></pre></div>
<p><img src="study_bias_doc_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n_bins =<span class="st"> </span><span class="dv">50</span>
n_proteins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(uniprot2pubmed_human_w)
n_proteins_per_bin =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(n_proteins<span class="op">/</span>n_bins)
bins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">50</span>, <span class="dt">each =</span> n_proteins_per_bin)
<span class="cf">if</span>(n_proteins <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)){
  bins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(bins, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">50</span>, n_proteins <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)))
} <span class="cf">else</span> <span class="cf">if</span>(n_proteins <span class="op">&lt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)){
  bins =<span class="st"> </span>bins[<span class="dv">1</span><span class="op">:</span>n_proteins]
}


uniprot2pubmed_human_w[, bins <span class="op">:</span><span class="er">=</span><span class="st"> </span>bins]
bin_lists =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/split">split</a></span>(uniprot2pubmed_human_w<span class="op">$</span>Entry, uniprot2pubmed_human_w<span class="op">$</span>bins)</code></pre></div>
</div>
<div id="second-we-need-to-count-interactions-within-and-between-bins" class="section level3">
<h3 class="hasAnchor">
<a href="#second-we-need-to-count-interactions-within-and-between-bins" class="anchor"></a>Second, we need to count interactions within and between bins</h3>
<p>This method shows the gist of the problem of study bias but needs further improvement.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">unique_interactions =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(human<span class="op">$</span>data[, .(IDs_interactor_A, IDs_interactor_B)])
N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bin_lists), <span class="cf">function</span>(bin_list1_N, unique_interactions, bin_lists){
    N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bin_lists), <span class="cf">function</span>(bin_list2_N, bin_list1_N, unique_interactions, bin_lists) {
      bin_list1 =<span class="st"> </span>bin_lists[[bin_list1_N]]
      bin_list2 =<span class="st"> </span>bin_lists[[bin_list2_N]]
      N_proteins =<span class="st"> </span><span class="kw">uniqueN</span>(unique_interactions[(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list1 <span class="op">&amp;</span>
<span class="st">                                     </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list2) <span class="op">|</span>
<span class="st">                                     </span>(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list2 <span class="op">&amp;</span>
<span class="st">                                     </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list1), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(IDs_interactor_A, IDs_interactor_B)])
      N_interactions =<span class="st"> </span>unique_interactions[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>((IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list1 <span class="op">&amp;</span>
<span class="st">                                     </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list2) <span class="op">|</span>
<span class="st">                                     </span>(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list2 <span class="op">&amp;</span>
<span class="st">                                     </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list1))]
      <span class="kw">data.table</span>(<span class="dt">N_interactions_per_protein =</span> N_interactions <span class="op">/</span><span class="st"> </span>N_proteins,
                 <span class="dt">x =</span> bin_list1_N, <span class="dt">y =</span> bin_list2_N,
                 <span class="dt">N_interactions =</span> N_interactions, <span class="dt">N_proteins =</span> N_proteins)
    }, bin_list1_N, unique_interactions, bin_lists)
    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(rbind, N_interaction_table)
}, unique_interactions, bin_lists)
N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(rbind, N_interaction_table)</code></pre></div>
</div>
<div id="third-we-need-to-plot-our-results" class="section level3">
<h3 class="hasAnchor">
<a href="#third-we-need-to-plot-our-results" class="anchor"></a>Third, we need to plot our results</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cols =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/colorRamp">colorRampPalette</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"white"</span>, <span class="st">"#95cbee"</span>, <span class="st">"#0099dc"</span>, <span class="st">"#4ab04a"</span>, <span class="st">"#ffd73e"</span>), <span class="dt">bias =</span> <span class="dv">2</span>)(<span class="dv">10</span>), <span class="co">#"#c9e2f6"</span>
            <span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/colorRamp">colorRampPalette</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"#eec73a"</span>, <span class="st">"#e29421"</span>, <span class="st">"#e29421"</span>, <span class="st">"#f05336"</span>,<span class="st">"#ce472e"</span>))(<span class="dv">20</span>))
<span class="kw">ggplot</span>(N_interaction_table, <span class="kw">aes</span>(x, y)) <span class="op">+</span>
<span class="st"> </span><span class="kw">geom_raster</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> N_interactions_per_protein)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">"rank by number of studies"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"rank by number of studies"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colours=</span>cols,
                         <span class="dt">na.value=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/rgb">rgb</a></span>(<span class="dv">246</span>, <span class="dv">246</span>, <span class="dv">246</span>, <span class="dt">max=</span><span class="dv">255</span>),
                         <span class="dt">guide=</span><span class="kw">guide_colourbar</span>(<span class="dt">ticks=</span>T, <span class="dt">nbin=</span><span class="dv">50</span>,
                                               <span class="dt">barheight=</span><span class="dv">15</span>, <span class="dt">label=</span>T))</code></pre></div>
<p><img src="study_bias_doc_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div id="finally-we-can-write-a-function-to-analyse-other-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#finally-we-can-write-a-function-to-analyse-other-datasets" class="anchor"></a>Finally, we can write a function to analyse other datasets</h3>
<p>First, let’s put what we just did into a function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">interactionBias =<span class="st"> </span><span class="cf">function</span>(ppi_network, uniprot2pubmed, <span class="dt">n_bins =</span> <span class="dv">50</span>){
  uniprot2pubmed =<span class="st"> </span><span class="kw">copy</span>(uniprot2pubmed)
  uniprot2pubmed =<span class="st"> </span>uniprot2pubmed[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(N_PMIDs, <span class="dt">decreasing =</span> T)]
  uniprot2pubmed =<span class="st"> </span>uniprot2pubmed[Entry <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="../../reference/extractInteractors.html">extractInteractors</a></span>(human)]
  
  n_proteins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(uniprot2pubmed)
  n_proteins_per_bin =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(n_proteins<span class="op">/</span>n_bins)
  bins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span><span class="op">:</span>n_bins, <span class="dt">each =</span> n_proteins_per_bin)
  <span class="cf">if</span>(n_proteins <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)){
    bins =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(bins, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(n_bins, n_proteins <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)))
  } <span class="cf">else</span> <span class="cf">if</span>(n_proteins <span class="op">&lt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bins)){
    bins =<span class="st"> </span>bins[<span class="dv">1</span><span class="op">:</span>n_proteins]
  }
  
  uniprot2pubmed[, bins <span class="op">:</span><span class="er">=</span><span class="st"> </span>bins]
  bin_lists =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/split">split</a></span>(uniprot2pubmed<span class="op">$</span>Entry, uniprot2pubmed<span class="op">$</span>bins)
  unique_interactions =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(ppi_network<span class="op">$</span>data[, .(IDs_interactor_A, IDs_interactor_B)])
  N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bin_lists), <span class="cf">function</span>(bin_list1_N, unique_interactions, bin_lists){
    N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bin_lists), <span class="cf">function</span>(bin_list2_N, bin_list1_N, unique_interactions, bin_lists) {
      bin_list1 =<span class="st"> </span>bin_lists[[bin_list1_N]]
      bin_list2 =<span class="st"> </span>bin_lists[[bin_list2_N]]
      N_proteins =<span class="st"> </span><span class="kw">uniqueN</span>(unique_interactions[(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list1 <span class="op">&amp;</span>
<span class="st">                                                  </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list2) <span class="op">|</span>
<span class="st">                                                 </span>(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list2 <span class="op">&amp;</span>
<span class="st">                                                    </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list1), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(IDs_interactor_A, IDs_interactor_B)])
      N_interactions =<span class="st"> </span>unique_interactions[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>((IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list1 <span class="op">&amp;</span>
<span class="st">                                                   </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list2) <span class="op">|</span>
<span class="st">                                                  </span>(IDs_interactor_A <span class="op">%in%</span><span class="st"> </span>bin_list2 <span class="op">&amp;</span>
<span class="st">                                                     </span>IDs_interactor_B <span class="op">%in%</span><span class="st"> </span>bin_list1))]
      <span class="kw">data.table</span>(<span class="dt">N_interactions_per_protein =</span> N_interactions <span class="op">/</span><span class="st"> </span>N_proteins,
                 <span class="dt">x =</span> bin_list1_N, <span class="dt">y =</span> bin_list2_N,
                 <span class="dt">N_interactions =</span> N_interactions, <span class="dt">N_proteins =</span> N_proteins)
    }, bin_list1_N, unique_interactions, bin_lists)
    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(rbind, N_interaction_table)
  }, unique_interactions, bin_lists)
  N_interaction_table =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(rbind, N_interaction_table)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(N_interaction_table)
}</code></pre></div>
</div>
<div id="you-can-see-that-marc-vidal-two-hybrid-and-mattias-mann-ap-ms-data-is-substantially-less-biased-by-protein-popularity-among-in-the-literature" class="section level2">
<h2 class="hasAnchor">
<a href="#you-can-see-that-marc-vidal-two-hybrid-and-mattias-mann-ap-ms-data-is-substantially-less-biased-by-protein-popularity-among-in-the-literature" class="anchor"></a>4.2 You can see that Marc Vidal two-hybrid and Mattias Mann AP-MS data is substantially less biased by protein popularity among in the literature</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">N_interaction_table[, title <span class="op">:</span><span class="er">=</span><span class="st"> "All IntAct data"</span>]
N_interaction_Mann =<span class="st"> </span><span class="kw">interactionBias</span>(<span class="dt">ppi_network =</span> Mann, <span class="dt">uniprot2pubmed =</span> uniprot2pubmed_human, <span class="dt">n_bins =</span> <span class="dv">50</span>)
N_interaction_Mann[, title <span class="op">:</span><span class="er">=</span><span class="st"> "Mattias Mann unbiased AP-MS dataset"</span>]

N_interaction_Vidal =<span class="st"> </span><span class="kw">interactionBias</span>(<span class="dt">ppi_network =</span> Vidal_all, <span class="dt">uniprot2pubmed =</span> uniprot2pubmed_human, <span class="dt">n_bins =</span> <span class="dv">50</span>)
N_interaction_Vidal[, title <span class="op">:</span><span class="er">=</span><span class="st"> "Marc Vidal unbiased two-hybrid dataset"</span>]

N_interaction_two_hybrid =<span class="st"> </span><span class="kw">interactionBias</span>(<span class="dt">ppi_network =</span> human_two_hybrid, <span class="dt">uniprot2pubmed =</span> uniprot2pubmed_human, <span class="dt">n_bins =</span> <span class="dv">50</span>)
N_interaction_two_hybrid[, title <span class="op">:</span><span class="er">=</span><span class="st"> "all two-hybrid data"</span>]

N_interaction_AP_MS =<span class="st"> </span><span class="kw">interactionBias</span>(<span class="dt">ppi_network =</span> human_AP_MS, <span class="dt">uniprot2pubmed =</span> uniprot2pubmed_human, <span class="dt">n_bins =</span> <span class="dv">50</span>)
N_interaction_AP_MS[, title <span class="op">:</span><span class="er">=</span><span class="st"> "all AP-MS data"</span>]

N_interaction =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(rbind, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(N_interaction_Mann, N_interaction_Vidal,
                                   N_interaction_two_hybrid, N_interaction_AP_MS, N_interaction_table))

<span class="kw">ggplot</span>(N_interaction, <span class="kw">aes</span>(x, y)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_raster</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> N_interactions_per_protein)) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"rank by number of studies"</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">"rank by number of studies"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">facet_wrap</span>( <span class="op">~</span><span class="st"> </span>title, <span class="dt">ncol =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colours=</span>cols,
                         <span class="dt">na.value=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/rgb">rgb</a></span>(<span class="dv">246</span>, <span class="dv">246</span>, <span class="dv">246</span>, <span class="dt">max=</span><span class="dv">255</span>),
                         <span class="dt">guide=</span><span class="kw">guide_colourbar</span>(<span class="dt">ticks=</span>T, <span class="dt">nbin=</span><span class="dv">50</span>,
                                               <span class="dt">barheight=</span><span class="dv">20</span>, <span class="dt">label=</span>T))</code></pre></div>
<p><img src="study_bias_doc_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
</div>
</div>
<div id="r-and-system-details" class="section level1">
<h1 class="hasAnchor">
<a href="#r-and-system-details" class="anchor"></a>R and system details</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Sys.time">Sys.Date</a></span>()</code></pre></div>
<pre><code>## [1] "2019-04-01"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">session_info</a></span>()</code></pre></div>
<pre><code>## ─ Session info ──────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.5.1 (2018-07-02)
##  os       macOS High Sierra 10.13.6   
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_GB.UTF-8                 
##  ctype    en_GB.UTF-8                 
##  tz       Europe/London               
##  date     2019-04-01                  
## 
## ─ Packages ──────────────────────────────────────────────────────────────
##  package       * version   date       lib source        
##  AnnotationDbi   1.44.0    2018-10-30 [1] Bioconductor  
##  assertthat      0.2.1     2019-03-21 [1] CRAN (R 3.5.1)
##  backports       1.1.3     2018-12-14 [1] CRAN (R 3.5.0)
##  Biobase         2.42.0    2018-10-30 [1] Bioconductor  
##  BiocGenerics  * 0.28.0    2018-10-30 [1] Bioconductor  
##  biomaRt       * 2.38.0    2018-10-30 [1] Bioconductor  
##  Biostrings      2.50.2    2019-01-03 [1] Bioconductor  
##  bit             1.1-14    2018-05-29 [1] CRAN (R 3.5.0)
##  bit64           0.9-7     2017-05-08 [1] CRAN (R 3.5.0)
##  bitops          1.0-6     2013-08-17 [1] CRAN (R 3.5.0)
##  blob            1.1.1     2018-03-25 [1] CRAN (R 3.5.0)
##  callr           3.2.0     2019-03-15 [1] CRAN (R 3.5.2)
##  cli             1.1.0     2019-03-19 [1] CRAN (R 3.5.2)
##  colorspace      1.4-1     2019-03-18 [1] CRAN (R 3.5.2)
##  commonmark      1.7       2018-12-01 [1] CRAN (R 3.5.0)
##  crayon          1.3.4     2017-09-16 [1] CRAN (R 3.5.0)
##  curl            3.3       2019-01-10 [1] CRAN (R 3.5.2)
##  data.table    * 1.12.0    2019-01-13 [1] CRAN (R 3.5.2)
##  DBI             1.0.0     2018-05-02 [1] CRAN (R 3.5.0)
##  desc            1.2.0     2018-05-01 [1] CRAN (R 3.5.0)
##  devtools        2.0.1     2018-10-26 [1] CRAN (R 3.5.1)
##  digest          0.6.18    2018-10-10 [1] CRAN (R 3.5.0)
##  downloader    * 0.4       2015-07-09 [1] CRAN (R 3.5.0)
##  dplyr           0.8.0.1   2019-02-15 [1] CRAN (R 3.5.2)
##  evaluate        0.13      2019-02-12 [1] CRAN (R 3.5.2)
##  fs              1.2.6     2018-08-23 [1] CRAN (R 3.5.0)
##  ggplot2       * 3.1.0     2018-10-25 [1] CRAN (R 3.5.1)
##  glue            1.3.1     2019-03-12 [1] CRAN (R 3.5.2)
##  gsubfn          0.7       2018-03-16 [1] CRAN (R 3.5.0)
##  gtable          0.3.0     2019-03-25 [1] CRAN (R 3.5.1)
##  hms             0.4.2     2018-03-10 [1] CRAN (R 3.5.0)
##  htmltools       0.3.6     2017-04-28 [1] CRAN (R 3.5.0)
##  httr          * 1.4.0     2018-12-11 [1] CRAN (R 3.5.0)
##  IRanges       * 2.16.0    2018-10-30 [1] Bioconductor  
##  jsonlite        1.6       2018-12-07 [1] CRAN (R 3.5.0)
##  knitr           1.22      2019-03-08 [1] CRAN (R 3.5.1)
##  labeling        0.3       2014-08-23 [1] CRAN (R 3.5.0)
##  lazyeval        0.2.2     2019-03-15 [1] CRAN (R 3.5.2)
##  magrittr        1.5       2014-11-22 [1] CRAN (R 3.5.0)
##  MASS            7.3-50    2018-04-30 [2] CRAN (R 3.5.1)
##  memoise         1.1.0     2017-04-21 [1] CRAN (R 3.5.0)
##  MItools       * 0.1.40    2018-11-20 [1] local         
##  munsell         0.5.0     2018-06-12 [1] CRAN (R 3.5.0)
##  ontologyIndex   2.5       2019-01-08 [1] CRAN (R 3.5.2)
##  pillar          1.3.1     2018-12-15 [1] CRAN (R 3.5.0)
##  pkgbuild        1.0.3     2019-03-20 [1] CRAN (R 3.5.1)
##  pkgconfig       2.0.2     2018-08-16 [1] CRAN (R 3.5.0)
##  pkgdown         1.3.0     2018-12-07 [1] CRAN (R 3.5.0)
##  pkgload         1.0.2     2018-10-29 [1] CRAN (R 3.5.0)
##  plyr          * 1.8.4     2016-06-08 [1] CRAN (R 3.5.0)
##  prettyunits     1.0.2     2015-07-13 [1] CRAN (R 3.5.0)
##  processx        3.3.0     2019-03-10 [1] CRAN (R 3.5.2)
##  progress        1.2.0     2018-06-14 [1] CRAN (R 3.5.0)
##  proto           1.0.0     2016-10-29 [1] CRAN (R 3.5.0)
##  ps              1.3.0     2018-12-21 [1] CRAN (R 3.5.0)
##  PSICQUIC      * 1.20.0    2018-10-30 [1] Bioconductor  
##  purrr           0.3.2     2019-03-15 [1] CRAN (R 3.5.2)
##  qvalue          2.14.1    2019-01-10 [1] Bioconductor  
##  R.methodsS3     1.7.1     2016-02-16 [1] CRAN (R 3.5.0)
##  R.oo            1.22.0    2018-04-22 [1] CRAN (R 3.5.0)
##  R.utils         2.8.0     2019-02-14 [1] CRAN (R 3.5.2)
##  R6              2.4.0     2019-02-14 [1] CRAN (R 3.5.2)
##  Rcpp            1.0.1     2019-03-17 [1] CRAN (R 3.5.2)
##  RCurl           1.95-4.12 2019-03-04 [1] CRAN (R 3.5.2)
##  remotes         2.0.2     2018-10-30 [1] CRAN (R 3.5.0)
##  reshape2        1.4.3     2017-12-11 [1] CRAN (R 3.5.0)
##  rlang           0.3.1     2019-01-08 [1] CRAN (R 3.5.2)
##  rmarkdown       1.11      2018-12-08 [1] CRAN (R 3.5.1)
##  roxygen2        6.1.1     2018-11-07 [1] CRAN (R 3.5.0)
##  rprojroot       1.3-2     2018-01-03 [1] CRAN (R 3.5.0)
##  RSQLite         2.1.1     2018-05-06 [1] CRAN (R 3.5.0)
##  rstudioapi      0.9.0     2019-01-09 [1] CRAN (R 3.5.2)
##  S4Vectors     * 0.20.1    2018-11-09 [1] Bioconductor  
##  scales          1.0.0     2018-08-09 [1] CRAN (R 3.5.0)
##  sessioninfo     1.1.1     2018-11-05 [1] CRAN (R 3.5.0)
##  stringi         1.4.3     2019-03-12 [1] CRAN (R 3.5.2)
##  stringr         1.4.0     2019-02-10 [1] CRAN (R 3.5.2)
##  testthat        2.0.1     2018-10-13 [1] CRAN (R 3.5.1)
##  tibble          2.1.1     2019-03-16 [1] CRAN (R 3.5.2)
##  tidyselect      0.2.5     2018-10-11 [1] CRAN (R 3.5.0)
##  usethis         1.4.0     2018-08-14 [1] CRAN (R 3.5.0)
##  withr           2.1.2     2018-03-15 [1] CRAN (R 3.5.0)
##  xfun            0.5       2019-02-20 [1] CRAN (R 3.5.2)
##  XML             3.98-1.19 2019-03-06 [1] CRAN (R 3.5.2)
##  xml2            1.2.0     2018-01-24 [1] CRAN (R 3.5.0)
##  XVector         0.22.0    2018-10-30 [1] Bioconductor  
##  yaml            2.2.0     2018-07-25 [1] CRAN (R 3.5.0)
##  zlibbioc        1.28.0    2018-10-30 [1] Bioconductor  
## 
## [1] /Users/vk7/Library/R/3.5/library
## [2] /Library/Frameworks/R.framework/Versions/3.5/Resources/library</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#outline">Outline</a></li>
      <li><a href="#brief-intro-into-r-markdown">0. Brief intro into R Markdown</a></li>
      <li><a href="#the-fundamentals-of-protein-protein-interactions-ppi-and-networks-lecture">1. The fundamentals of protein-protein interactions (PPI) and networks (lecture)</a></li>
      <li>
<a href="#getting-ppi-data-into-r-from-intact-using-psicquic">2. Getting PPI data into R (from IntAct, using PSICQUIC)</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-ppi-data-into-r-from-intact">2.1 Getting PPI data into R from IntAct</a></li>
      <li><a href="#getting-ppi-data-from-non-imex-databases-into-r-using-psicquic">2.2 Getting PPI data from non-IMEx databases into R using PSICQUIC</a></li>
      <li><a href="#advanced-search">2.3 Advanced search</a></li>
      </ul>
</li>
      <li>
<a href="#finding-publication-pubmed-ids-for-every-protein">3. Finding publication (Pubmed) IDs for every protein</a><ul class="nav nav-pills nav-stacked">
<li><a href="#retrive-uniprotkb-references-using-uniprot-rest-api">3.1 Retrive UniProtKB references using Uniprot REST API</a></li>
      </ul>
</li>
      <li>
<a href="#evaluating-bias-of-protein-popularity-in-the-literature">4. Evaluating bias of protein popularity in the literature</a><ul class="nav nav-pills nav-stacked">
<li><a href="#you-can-see-that-marc-vidal-two-hybrid-and-mattias-mann-ap-ms-data-is-substantially-less-biased-by-protein-popularity-among-in-the-literature">4.2 You can see that Marc Vidal two-hybrid and Mattias Mann AP-MS data is substantially less biased by protein popularity among in the literature</a></li>
      </ul>
</li>
      <li><a href="#r-and-system-details">R and system details</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Vitalii Kleshchevnikov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
