<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Molecular interaction data tools: retrieving data from IntAct and most other databases using PSICQUIC service â€¢ MItools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Molecular interaction data tools: retrieving data from IntAct and most other databases using PSICQUIC service">
<meta property="og:description" content="Package contains high-level functions for retrieving molecular interaction data from most public databases using PSICQUIC service (http://www.ebi.ac.uk/Tools/webservices/psicquic/view/main.xhtml) and manipulating that data. Useful features include: retrieve full interactome of a given species, retrieve all interactions between two taxonomy groups (taxonomy structure-aware search), query molecular interaction databases using MIQL, subset interaction data by list of protein/gene identifiers, interaction detection method, publication (PMIDs). Depends on PSICQUIC, data.table.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MItools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.41</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/articles/Full_interactomes_interspecies.html">Getting started: full interactomes and interspecies interactions</a>
    </li>
    <li>
      <a href="articles/articles/MItools_vs_igraph.html">MItools vs igraph: calculating degree of a protein</a>
    </li>
    <li>
      <a href="articles/articles/study_bias_doc.html">Study bias in the protein network</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="r-package---mitools" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#r-package---mitools" class="anchor"></a>R Package - MItools</h1></div>
<div id="molecular-interaction-data-tools-retrieving-data-from-intact-and-most-other-databases-using-psicquic-service" class="section level2">
<h2 class="hasAnchor">
<a href="#molecular-interaction-data-tools-retrieving-data-from-intact-and-most-other-databases-using-psicquic-service" class="anchor"></a>Molecular interaction data tools: retrieving data from IntAct and most other databases using PSICQUIC service</h2>
<p>Package contains high-level functions for retrieving molecular interaction data from most public databases using PSICQUIC service (<a href="http://www.ebi.ac.uk/Tools/webservices/psicquic/view/main.xhtml" class="uri">http://www.ebi.ac.uk/Tools/webservices/psicquic/view/main.xhtml</a>) and manipulating that data. Useful features include: retrieve full interactome of a given species, retrieve all interactions between two taxonomy groups (taxonomy structure-aware search), query molecular interaction databases using MIQL, subset interaction data by list of protein/gene identifiers, interaction detection method, publication (PMIDs). Depends on PSICQUIC, data.table.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Install R MItools package</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>) <span class="co"># for installing BioConductor dependencies</span>
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"vitkl/MItools"</span>, <span class="dt">dependencies =</span> T)</code></pre></div>
</div>
<div id="basic-use" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-use" class="anchor"></a>Basic use</h2>
<p>Load all interactions between human proteins in IntAct database.</p>
<p>Follow this <a href="https://vitkl.github.io/MItools/articles/articles/Full_interactomes_interspecies.html">example</a> for more details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load package</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MItools)

<span class="co"># load all interactions from IntAct FTP storage (https://www.ebi.ac.uk/intact/downloads)</span>
human =<span class="st"> </span><span class="kw"><a href="reference/fullInteractome.html">fullInteractome</a></span>(<span class="dt">taxid =</span> <span class="dv">9606</span>, <span class="dt">database =</span> <span class="st">"IntActFTP"</span>,
                        <span class="dt">format =</span> <span class="st">"tab27"</span>,
                        <span class="dt">clean =</span> <span class="ot">TRUE</span>, <span class="co"># parse into usable format (takes 5-10 minutes)</span>
                        <span class="dt">protein_only =</span> <span class="ot">TRUE</span>, <span class="co"># filter protein interactions</span>
                        <span class="dt">directory =</span> <span class="ot">NULL</span>) <span class="co"># keep data files inside R library, </span>
                                          <span class="co"># or specify your directory </span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># protein interactions lack directionality &amp; are detected in multiple studies</span>
<span class="co"># so find the number of unique interactions</span>
<span class="kw"><a href="reference/extractInteractors.html">uniqueNinteractions</a></span>(human)
<span class="kw"><a href="reference/extractInteractors.html">uniqueNinteractors</a></span>(human)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset by a list of proteins</span>
mediator_complex_proteins =<span class="st"> </span><span class="kw">fread</span>(<span class="st">"https://www.uniprot.org/uniprot/?query=GO:0016592%20AND%20taxonomy:9606&amp;format=tab&amp;columns=id"</span>)
mediator_complex =<span class="st"> </span><span class="kw"><a href="reference/subsetMITABbyID.html">subsetMITABbyID</a></span>(human, <span class="dt">ID_seed =</span> mediator_complex_proteins<span class="op">$</span>Entry,
                                   <span class="dt">within_seed =</span> T, <span class="dt">only_seed2nonseed =</span> F)
<span class="kw"><a href="reference/extractInteractors.html">uniqueNinteractions</a></span>(mediator_complex)
<span class="kw"><a href="reference/extractInteractors.html">uniqueNinteractors</a></span>(mediator_complex)</code></pre></div>
<p>Load interactions for one protein identified with a specific method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Query for interactions of bacterial RNA polymerase sigma factor SigA </span>
<span class="co"># identified using two-hybrid methods inall imex databases </span>
<span class="co"># (result identical to IntActFTP but only requested interactions are downloaded)</span>
inter =<span class="st"> </span><span class="kw"><a href="reference/queryPSICQUICrlib.html">queryPSICQUICrlib</a></span>(<span class="dt">query =</span> <span class="st">"id:P74565 AND detmethod:</span><span class="ch">\"</span><span class="st">MI:0018</span><span class="ch">\"</span><span class="st">"</span>,
                  <span class="dt">format =</span> <span class="st">"tab25"</span>,
                  <span class="dt">database =</span> <span class="st">"imex"</span>,
                  <span class="dt">directory =</span> <span class="st">"./data_files/"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The data returned by queryPSICQUICrlib constains auxillary information </span>
<span class="co"># that is not necessary for most analysis. Let's clean the data.</span>
inter =<span class="st"> </span><span class="kw"><a href="reference/cleanMITAB.html">cleanMITAB</a></span>(inter)
inter<span class="op">$</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<p>Beware of <a href="https://vitkl.github.io/MItools/articles/articles/study_bias_doc.html">study bias</a> (ascertainment bias): in aggregated data better studied proteins have more interactions (not a problem for unbiased proteome-wide interaction screens).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Report a bug at <br><a href="https://github.com/vitkl/MItools/issues">https://â€‹github.com/â€‹vitkl/â€‹MItools/â€‹issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>Apache License 2.0</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Vitalii Kleshchevnikov <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by Vitalii Kleshchevnikov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
